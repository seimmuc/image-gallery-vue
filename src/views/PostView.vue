<script setup lang="ts">
import { computed } from '@vue/reactivity';
import { ref, watchEffect, type Ref } from 'vue';
import { useRoute, useRouter } from 'vue-router';
import Post from '../components/Post.vue';
import { type IPost, type IPool, fetchPost } from '../api';

const route = useRoute();
const router = useRouter();
const post: Ref<IPost | null> = ref(null);

// loads new post data when params.id changes
watchEffect(
  async () => {
    const id = parseInt(route.params.id as string);
    if (!Number.isSafeInteger(id))
      return;
    try {
      post.value = await fetchPost(id);
    } catch (SyntaxError) {
      post.value = null;
    }
  }
);

// returns current pool as specified by url query, if it's present on the post object
const currentPool = computed((): IPool | null => {
  const poolName = route.query.pool;
  if (post.value === null || poolName === undefined || !post.value.pools)
    return null;
  const pool = post.value.pools.find(p => p.name == poolName);
  return pool? pool : null;
});

// previous/next post computed properties and navigation functions
const prevPostId = computed((): number | undefined => currentPool.value ? currentPool.value.previd : undefined)
const nextPostId = computed((): number | undefined => currentPool.value ? currentPool.value.nextid : undefined)
function gotoPrev(e: MouseEvent) {
  if (!prevPostId.value)
    return;
  router.push({name: 'post', params: {id: prevPostId.value}, query: route.query});
}
function gotoNext(e: MouseEvent) {
  if (!nextPostId.value)
    return;
  router.push({name: 'post', params: {id: nextPostId.value}, query: route.query});
}
</script>

<template>
  <main class="post-view">
    <Post :post="post" :goto-prev="prevPostId? gotoPrev : null" :goto-next="nextPostId? gotoNext : null" />
  </main>
</template>

<style scoped>
.image-container {
  display: flex;
}

.navigate {
  overflow: hidden;
  white-space: nowrap;
}
</style>
