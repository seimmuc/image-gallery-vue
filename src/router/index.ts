import { createRouter, createWebHistory } from 'vue-router'
import HomeView from '../views/HomeView.vue'
import PostView from '../views/PostView.vue'

const routes = [
  {
    path: '/',
    name: 'home',
    component: HomeView
  },
  {
    path: '/post/:id(\\d+)',
    name: 'post',
    component: PostView
  },
  {
    path: '/about',
    name: 'about',
    component: () => import('../views/AboutView.vue')
  }
]

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: routes
})

export {routes}
export default router
