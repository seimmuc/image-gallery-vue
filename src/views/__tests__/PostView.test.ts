import { describe, it, expect, vi, beforeEach } from 'vitest'

import { flushPromises, mount } from '@vue/test-utils'
import { routes } from '../../router/index';
import PostView from '../PostView.vue';
import { createRouter, createWebHistory, type Router } from 'vue-router';
import createFetchMock from 'vitest-fetch-mock';
import postData from './PostView.data.json';


const mountGlobal = { stubs: ['FontAwesomeIcon'] };
const fetchMock = createFetchMock(vi);
fetchMock.enableMocks();

// mock posts API /api/post/<id>
const postUrlRegex = /\/api\/post\/(\d+)\/?(?:\?.*)?/
fetchMock.doMock((req) => {
  const match = req.url.match(postUrlRegex)
  if (!match || !match[1] || !(match[1] in postData))
    return { status: 404, body: '{}' };
  return { status: 200, body: JSON.stringify(postData[<keyof typeof postData> match[1]]) };
});

describe('PostView', () => {

  let router: Router;
  beforeEach(async () => {
    router = createRouter({
      history: createWebHistory(import.meta.env.BASE_URL),
      routes: routes,
    });
    // router.push('/');
    // await router.isReady();
  });

  it('first post', async () => {
    router.push({name: 'post', params: {id: 1}});
    await router.isReady();
    const wrapper = mount(PostView, {global: {...mountGlobal, plugins: [router]}});
    wrapper.vm.$router.push({name: 'post', params: {id: 1}});
    await flushPromises();
    expect(wrapper.vm.$router).toBe(router);
    
    expect(wrapper.html()).toMatchSnapshot();
  });
});
