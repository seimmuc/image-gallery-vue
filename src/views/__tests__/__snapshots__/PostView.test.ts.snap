// Vitest Snapshot v1

exports[`PostView > first post 1`] = `
"<main class=\\"post-view\\" data-v-5298551c=\\"\\">
  <div class=\\"post-container\\" data-v-a0662bbd=\\"\\" data-v-5298551c=\\"\\">
    <h1 data-v-a0662bbd=\\"\\">First</h1>
    <div class=\\"image-container\\" data-v-a0662bbd=\\"\\"><img class=\\"post-image\\" data-v-a0662bbd=\\"\\" src=\\"/images/first_image.png\\" alt=\\"first post\\">
      <!--v-if-->
      <!--v-if-->
    </div>
    <p data-v-a0662bbd=\\"\\">first post</p>
  </div>
</main>"
`;
